### PR TITLE
Add HTTP compression at origin (tower-http + Traefik)

### PR DESCRIPTION
**Pull request by ITC-MSM**
## Summary

Addresses the two low-effort levers identified in #7968's compression audit:

- **Origin HTTP compression**: adds `tower_http::CompressionLayer` (gzip) to
  the axum router in `crates/phase-server`, so REST/JSON responses (health
  check, P2P draft backup API, `/admin/*`) are compressed at origin
  regardless of which edge (or lack of one) sits in front of the server.
- **Traefik k8s path**: adds a `compress` Traefik `Middleware` CRD to
  `deploy/helm/phase-server`, wired into the same middleware chain as the
  existing `ratelimit`/`inflight`/`headers`/`backup-ratelimit`/
  `backup-buffering` middlewares, on both the `Ingress` and `IngressRoute`
  templates.

Lever #3 (WebSocket permessage-deflate / RFC 7692) is out of scope — it
requires a WS-stack change (tungstenite 0.29 has no implementation) and real
maintainer buy-in, per the issue.

Lever #4 (reusing the WebRTC P2P DataChannel's gzip envelope —
`client/src/network/protocol.ts:370-406` — on the client<->server WS adapter)
was evaluated but **not implemented here**: the client<->server transport
currently assumes plain-JSON WebSocket *text* frames end-to-end (client
`JSON.stringify`/`JSON.parse`, server `axum::extract::ws::Message` dispatch,
every wire guard in `server-core`), and this protocol already has real
version-skew handling (`PROTOCOL_VERSION`/`MIN_SUPPORTED_PROTOCOL`) because a
connected client and server are not guaranteed to be on the same build.
Introducing a binary gzip envelope safely needs a capability negotiation (or
protocol version bump) so a server never sends a compressed frame to a client
that can't decode it, plus correct behavior for mixed-version broadcast
fan-out during rolling deploys. That's genuine protocol design work, not a
config change, so it's filed separately: phase-rs/phase#8274.

## Root cause and fix

axum 0.8.9's WS upgrade never negotiates `Sec-WebSocket-Extensions:
permessage-deflate` (tungstenite 0.29 has no RFC 7692 support), so WS traffic
stays uncompressed regardless of what's done at the HTTP layer — that part is
unfixable without a WS-stack change (lever #3, explicitly out of scope). But
`crates/phase-server`'s HTTP responses (health check, the P2P draft backup
API, `/admin/*`) had no compression at origin at all: `tower-http` was built
with only the `cors` feature, and nothing in `crates/` added a
`CompressionLayer`. On the k8s deploy path, Traefik terminated TLS
unconditionally with no `compress` middleware in the rendered chain, so those
same responses stayed uncompressed on that hop too, unless an edge (e.g.
Cloudflare) happened to compress them downstream. Both origins now compress
their own HTTP responses independent of what's in front of them.

### `crates/phase-server`

- `Cargo.toml`: added the `compression-gzip` feature to the existing
  `tower-http` dependency (kept minimal — no brotli/zstd/deflate — consistent
  with the codebase's existing gzip precedent in the P2P wire format).
- `src/main.rs`: added a `build_router` function that assembles the router as
  two pieces merged together — a `/ws`-only sub-router with no compression,
  and every other route (`/health`, `/p2p-draft-backup*`, and `/admin/*` when
  enabled) wrapped in `CompressionLayer::new()`. `/ws` is kept off the
  compressed sub-router because its `101 Switching Protocols` response
  carries no body to compress, and axum's upgrade hands the connection off
  via `hyper`'s `OnUpgrade` extension rather than through the response body a
  compressing layer wraps — keeping the layer off that route entirely removes
  any dependence on that implementation detail continuing to hold. `main()`
  now calls `build_router(app_state.clone(), cors, admin_token.as_deref())`
  instead of assembling the router inline.
- New `compression_tests` module (3 tests) exercising `build_router`
  directly:
  - `health_response_is_uncompressed_regardless_of_accept_encoding` — pins
    the (correct) behavior that `CompressionLayer`'s default predicate skips
    `/health`'s 2-byte body even with `Accept-Encoding: gzip` sent, since
    compressing it would grow rather than shrink the response.
  - `large_json_response_is_gzip_compressed_when_accepted` — seeds a P2P
    draft backup snapshot large enough to clear the predicate's minimum size,
    fetches it with `Accept-Encoding: gzip`, and asserts a
    `Content-Encoding: gzip` header, a body starting with the gzip magic
    bytes, and a compressed body strictly smaller than the uncompressed JSON.
  - `ws_upgrade_still_succeeds_through_the_compressed_router` — connects to
    `/ws` (via `tokio_tungstenite::connect_async`) through the very `Router`
    that also carries `CompressionLayer`, guarding against the regression
    this fix must not introduce.

### `deploy/helm/phase-server`

- `values.yaml`: new `traefik.middlewares.compress.minResponseBodyBytes`
  (default `1024`, matching Traefik's own compress-middleware default),
  documented as safe on `/ws` for the same reason as above (no body to
  compress on the upgrade response, and compression only ever touches
  ordinary HTTP responses).
- `templates/middlewares.yaml`: new `{{ $fullname }}-compress` `Middleware`
  CRD (`spec.compress.minResponseBodyBytes`), matching the syntax/labels
  pattern of its five siblings.
- `templates/_helpers.tpl`: wired the new middleware into
  `phase-server.commonMiddlewares` (used by the annotation-based `Ingress`)
  and `phase-server.middlewareRefs` (used by the native `IngressRoute`), so
  it's applied everywhere the other three "applied to every routed path"
  middlewares are — including the `/p2p-draft-backup*` routes via
  `commonMiddlewares`' inclusion in `$backupMw`.

## Files changed

- `crates/phase-server/Cargo.toml`
- `crates/phase-server/src/main.rs`
- `deploy/helm/phase-server/values.yaml`
- `deploy/helm/phase-server/templates/middlewares.yaml`
- `deploy/helm/phase-server/templates/_helpers.tpl`
- `Cargo.lock`

## Verification

- `cargo fmt --all` — clean.
- `cargo build -p phase-server` — clean (Tilt is not running in this
  environment per the task; ran directly).
- `cargo clippy -p phase-server --all-targets -- -D warnings` — clean, no
  warnings.
- `cargo test -p phase-server` — **168 passed, 0 failed**, including the 3
  new `compression_tests` above and the full pre-existing suite (admin auth,
  P2P backup, WS handshake/mode-gate/metrics/persistence tests all still
  pass unchanged, confirming the router refactor didn't regress anything).
- Manual read-through of the rendered Helm template syntax (no local k8s
  cluster to render/apply against): the new `compress` Middleware mirrors its
  five siblings exactly (same `apiVersion`/`kind`/labels/`$fullname` pattern,
  `int64` cast on the numeric field), and is appended in both
  `commonMiddlewares` and `middlewareRefs` at the same point the other three
  "applied to every route" middlewares are.
- Gate A (`./scripts/check-parser-combinators.sh upstream/main`): `Gate G
  PASS` / `Gate A PASS head=4c97ba00eb210e842b8f21d19de4d4b39497d2dd
  base=1fc3a12e467c554710b235fab9c0e4c43abd6bfa` — expected, no parser code
  touched.
- Scope check: `git diff --stat upstream/main...HEAD` shows only the 6 files
  above.

## Follow-up

Addresses the low-effort levers in #7968 (CompressionLayer + Traefik
compress middleware); WS-level compression (permessage-deflate, lever #3)
remains out of reach without a WS-stack change, and reusing the P2P gzip
wire envelope on the client<->server WS adapter (lever #4) is tracked
separately in phase-rs/phase#8274 since it needs real protocol-versioning
design, not a config change.

---

Model: claude-sonnet-5
Tier: Frontier

## Summary by CodeRabbit

* **New Features**
  * Added gzip compression for sufficiently large HTTP responses.
  * Configured Helm deployments to compress standard HTTP routes while excluding WebSocket connections.
  * Added dedicated WebSocket routing to maintain reliable upgrade handling.

* **Bug Fixes**
  * Improved route matching so WebSocket traffic is consistently directed to its dedicated route.

* **Tests**
  * Added coverage for response compression thresholds, WebSocket upgrades, and rendered routing configuration.

### CODE REVIEW

File: deploy/helm/phase-server/templates/_helpers.tpl [hunk 1]
```diff
@@ -77,6 +77,7 @@ app.kubernetes.io/instance: {{ .Release.Name }}
 {{- $list = append $list (include "phase-server.crdRef" (dict "ctx" . "suffix" "ratelimit")) -}}
 {{- $list = append $list (include "phase-server.crdRef" (dict "ctx" . "suffix" "inflight")) -}}
 {{- $list = append $list (include "phase-server.crdRef" (dict "ctx" . "suffix" "headers")) -}}
+{{- $list = append $list (include "phase-server.crdRef" (dict "ctx" . "suffix" "compress")) -}}
```
[the remark this answers is not in the document]
Comment (ITC-MSM) [reply deploy/helm/phase-server/templates/_helpers.tpl]: Fixed in 1fbc4a812. `phase-server.commonMiddlewares` and `phase-server.middlewareRefs` now take `dict "ctx" $ "excludeCompress" true|false`.

- `ingress.yaml`: `/ws` is now its own Ingress resource (`<fullname>-ws`, previously merged with `/health`) built with `excludeCompress=true`; `/health` keeps its own Ingress with `compress`; `/p2p-draft-backup` is untouched.
- `ingressroute.yaml`: since `IngressRoute` here matches by `Host()` only (no path-based route split like the plain Ingress), I added a second, more specific `routes[]` rule per host — `Host(...) && PathPrefix("/ws")` with the compress-free middleware list and `priority: 100` — ahead of the existing general `Host(...)` rule (`priority: 1`, unchanged middleware list). Applied to both the entry route and every per-ordinal route; the sticky-cookie config on the entry route is preserved identically on both of its rules.

This mirrors the split `build_router` already makes in `crates/phase-server/src/main.rs` (`CompressionLayer` kept off `/ws`'s sub-router) for the same reason: a WS upgrade's 101 response has no body to compress, so running Traefik's compress middleware against it is pure overhead.

Verified with `helm template`/`helm lint` (v3.16.3, matching CI) against both default and scale-out values — `/ws` never carries the `-compress` middleware in either rendering path, while `/health`, `/p2p-draft-backup`, and the general routes still do. Re-ran the full `helm-chart.yml` assertion suite locally; all green, including the `kind: IngressRoute` count invariant (unaffected — this adds a route rule, not a new resource).

**Review by matthewevans — CHANGES_REQUESTED** [review]
## Current-head changes requested

Reviewed at `1fbc4a812beed575cb2cd21bf44d9cd7105afcea`.

**[MED] Add discriminating chart-render assertions for the compression boundary.** The exact behavior repaired in this head is encoded in `deploy/helm/phase-server/templates/ingress.yaml:13-28` and `deploy/helm/phase-server/templates/ingressroute.yaml:46-83,108-127`: `/ws` must omit `-compress`, while `/health`, backup, and ordinary/ordinal HTTP routes retain it. But the repository's Helm CI only establishes that the default render is nonempty (`.github/workflows/helm-chart.yml:42-46`) and checks scale-out resource kinds (`:73-85`); it does not assert either side of that routing contract. The earlier CodeRabbit `/ws` finding is resolved in this head, but this gap would let the same deployment regression return with green CI.

Please add focused default and scale-out rendered-manifest assertions that prove `/ws` is compression-free and the corresponding non-WebSocket routes carry the compression middleware, including an ordinal route. Then request re-review.

The type is `bug`; this PR is not approved or enqueued.

**Comment by ITC-MSM** [comment]
Addressed the current-head chart-test request in e285802c. The Helm workflow now asserts the rendered compression boundary for both deployment shapes:\n\n- default Ingress: /health and /p2p-draft-backup retain -compress@kubernetescrd; /ws omits it\n- scale-out IngressRoute: both the entry route and representative ordinal phase-server-0 have a compression-free /ws rule and a compressed ordinary HTTP rule\n\nValidated locally with Helm v3.16.3: chart lint passed, default and scale-out manifests rendered, and the exact awk/grep assertions passed against those manifests. Please re-review the updated head.

**Review by matthewevans — CHANGES_REQUESTED** [review]
## Current-head changes requested — Security/Sanity hard stop

Reviewed at `ee285802ce56c5745f4c6b76f0c8c909213ebe35`.

**[HIGH] External contributor changes CI workflow configuration.** Evidence: the exact current-head diff modifies `.github/workflows/helm-chart.yml` and adds a 43-line shell assertion step at lines 87–129. The contributor-handler Security/Sanity pre-check treats any external PR modification under `.github/workflows/` as a hard stop, independent of whether the assertions themselves appear reasonable.

Please split the workflow change out and obtain explicit maintainer review through the appropriate trusted path. This handler will not patch, approve, label, enqueue, or otherwise progress this PR while the workflow edit remains in the contributor diff.

The existing `bug` label is unchanged; auto-merge and merge-queue entry are absent.

**Comment by ITC-MSM** [comment]
Addressed the Security/Sanity hard stop at head fbc5f178c.

The contributor PR no longer modifies any file under .github/workflows. I moved the rendered-manifest assertions into the executable chart-owned script deploy/helm/phase-server/tests/assert-compression-boundary.sh, keeping the default and scale-out coverage requested in the earlier review:
- default /health and /p2p-draft-backup are compressed; /ws is not
- scale-out entry and ordinal phase-server-0 routes keep /ws uncompressed and ordinary HTTP compressed

This cleanly separates the contributor-owned test logic from trusted CI wiring. A maintainer can invoke the script from the existing Helm workflow through the trusted path when appropriate.

Fresh validation:
- helm lint with v3.16.3: passed
- default and scale-out renders: passed
- chart-owned boundary script against both renders: passed
- bash syntax check and git diff --check: passed

Please re-review the contributor-safe head. No workflow file remains in the PR diff.

**Comment by matthewevans** [comment]
**Maintainer hold for current head `1c335312e5bee5c33fa00b83fdfe7040b3c46602`.**

I added the missing trusted CI invocation of `deploy/helm/phase-server/tests/assert-compression-boundary.sh` after the default and scale-out renders, then brought the branch current with `main`. The script syntax and diff check pass; no direct build was run under the Tilt-first policy.

Required Helm/security/review checks have restarted, and the only parse-diff receipt names old head `ee285802ce56c5745f4c6b76f0c8c909213ebe35`. Recheck after terminal required CI and an exact-head parse receipt; no contributor action is requested.

**Comment by lgray** [comment]
:robot: _AI text below_ :robot:

Deep review of head 1c335312e (requested one-off; comment-only). **Verdict: the fix is sound — no blockers.** The origin `CompressionLayer` split, the Traefik middleware wiring, the chart test, and the PR body's technical premises all check out against the locked dependency sources, rendered manifests, and a local run of the new tests. One config-dependent Traefik finding worth a decision before merge, then FYIs.

**Findings (most severe first)**

1. **[Minor, config-dependent] Explicit `priority: 1` on the general `Host()` rules lowers them below every length-defaulted router on the entrypoint** — `deploy/helm/phase-server/templates/ingressroute.yaml` (both the entry route and each ordinal route). Traefik router priority is entrypoint-global and defaults to the rule's character length; before this PR the plain `Host(...)` router had default priority ≈26, now it is 1. In a shared Traefik where any other router's rule also matches these hosts — e.g. a cluster-wide ``PathPrefix(`/`)`` fallback/404 backend (default priority ≥15) — that router now outranks the chart's HTTP routes and silently swallows all non-`/ws` traffic that previously routed here. Within the chart nothing collides (verified: no other router/priority in `deploy/`), so this only bites shared-ingress clusters. Cheapest fix: drop both explicit priorities — the `&& PathPrefix(`/ws`)` rule string is strictly longer than the plain `Host()` rule, so Traefik's default ordering already prefers it (the template comment says as much); alternatively keep explicit values but well above typical rule lengths (e.g. 1000/900).

2. **[FYI] `.github/workflows/helm-chart.yml` is a `[hard_stops]` path per `.agents/pr-review-policy.toml`.** The only workflow change in this PR is the maintainer-authored commit 1c335312e (matthewevans), so the trusted-path requirement is satisfied; noting so the record is explicit. `crates/phase-server/src/main.rs` is separately on the `[architecture_scope]` list (mode = review), i.e. this PR requires a maintainer's full implementation review — which is what this is.

3. **[FYI] BREACH surface considered, none found.** `/admin/*` and backup responses are now gzipped over TLS. No response mixes a secret with attacker-reflected data: admin auth is a request-side bearer token never echoed, and `GET /p2p-draft-backup/{code}` returns only the stored blob to the peer that can name code + `host_peer_id`. No compression-oracle pairing identified; nothing to change.

**Against #7968**

The issue's lever list is delivered exactly as written: lever 1 (`tower_http::CompressionLayer` at origin, "one feature flag + one layer") and lever 2 (Traefik `compress` middleware, "config-only") are both implemented. The lever-3 deferral matches the issue's own text — "WS via RFC 7692 requires a library that implements it — current tungstenite 0.29 does not" — and I confirmed tungstenite 0.29.0's only permessage-deflate mentions are a header-parsing test fixture (`src/handshake/headers.rs`), and axum 0.8.9's `extract/ws.rs` never touches `Sec-WebSocket-Extensions`. Lever 4 is filed as #8274 (open); its deferral premise holds: the WS adapter is plain-JSON text frames end-to-end and `PROTOCOL_VERSION`/`MIN_SUPPORTED_PROTOCOL` (`crates/server-core/src/protocol.rs:23-27`) confirm real version-skew handling exists, so a binary envelope does need negotiation. The PR body carries no closing keyword, so merging will not auto-close #7968 — correct, since the issue's WSS half (its title's first clause) remains unaddressed by design; #7968 should stay open re-scoped to levers 3/4 after merge.

**Verified clean (one line each)**

- `build_router` layering: compression is applied only to the HTTP sub-router before merge, CORS stays outermost for all routes exactly as before; `/ws` never passes through `CompressionLayer` (crates/phase-server/src/main.rs).
- All three new `compression_tests` pass locally (small-body skip, gzip magic bytes + smaller-than-plaintext on the backup API, WS handshake through the merged router); the tests are in the bin target and the bin has no `test = false`, so they actually run.
- tower-http 0.6.10 default predicate verified in source: `SizeAbove(32)` + not-for grpc/image/SSE — `application/json` is compressed, `/health`'s 2-byte body is skipped, and the layer appends `Vary: accept-encoding` (compression/future.rs), so no cache poisoning.
- Feature delta is minimal: only `compression-gzip` added in `crates/phase-server/Cargo.toml`; Cargo.lock delta is 2 lines (`flate2`, `memchr` under `compression-codecs`); no new dependencies (`tokio-tungstenite` was already a dev-dep).
- No double compression: Traefik's compress middleware documented to skip responses already bearing `Content-Encoding` (traefik.io compress middleware reference), and origin gzip always sets it first; `minResponseBodyBytes: 1024` matches Traefik's documented default.
- Middleware chain position: `compress` appended last on both the annotation (`commonMiddlewares`) and CRD-ref (`middlewareRefs`) paths, so ratelimit/inflight reject on the request phase before compression is ever engaged and response compression sits closest to the service; existing chains are otherwise byte-identical (rendered and diffed both deployment shapes).
- Rendered boundary: default Ingress render gives `/health` and `/p2p-draft-backup` the `-compress@kubernetescrd` ref and `/ws` (its own `-ws` Ingress) none; scale-out render gives every ``PathPrefix(`/ws`)`` rule a compress-free list and every general rule the compress ref, sticky-cookie config preserved identically on both entry rules.
- `assert-compression-boundary.sh` is non-vacuous and discriminating: mutating a render to drop compress from the HTTP chain fails it, leaking compress onto a `/ws` rule fails it, and the unmutated renders pass; CI's "assert compression boundary" step ran and passed at this head.
- Engine purity: the change is pure transport/deploy config — no game logic touched, no engine surface moved.

**Comment by lgray** [comment]
:robot: _AI text below_ :robot:

Supplement to my review comment ([#issuecomment-5502787928](https://github.com/phase-rs/phase/pull/8275#issuecomment-5502787928)), answering a follow-up question: **does this PR compress the cluster-internal pod→Traefik hop, or only the Traefik→client edge?** Answer, measured: **the internal hop is compressed too — the origin does the work and Traefik just forwards.**

**Measured (live Traefik v3.5 + compress middleware with this chart's `minResponseBodyBytes: 1024`, upstream echoing received headers):**

- Traefik forwards the client's `Accept-Encoding` upstream verbatim — the compress middleware only reads it (Traefik v3.5 `pkg/middlewares/compress/compress.go`, `ServeHTTP`; the gzhttp gzip wrapper likewise). So the PR's origin `CompressionLayer` sees the header first and gzips eligible responses (≥32B, per the PR's own tests).
- The already-encoded response passes Traefik unchanged: exactly one `Content-Encoding: gzip`, one gunzip pass yields the payload. gzhttp compresses only when the response's `Content-Encoding` is empty (klauspost/compress `gzhttp/compress.go:166`), matching the Traefik doc condition cited in the main review. No double compression; the edge is not doing the heavy work.
- When the client sends **no** `Accept-Encoding`, Traefik's Go transport adds its own `Accept-Encoding: gzip` upstream and transparently decodes for the client — so the internal hop carries gzip even for non-accepting clients.

**Still uncompressed on the internal hop (all outside this PR's scope):**

- `/ws` frames, both directions — the realtime game/lobby/draft channel, structurally the bulk of internal bandwidth (levers 3/4 of #7968, deferred; lever 4 tracked in #8274).
- Request bodies everywhere (`CompressionLayer` is response-only), including the ≤1MiB per-pick `/p2p-draft-backup` POST snapshots — pre-existing, noted in #7968 itself.
- Sub-32B responses (`/health` — correct, compression would grow them).
- `/metrics`: served from the separate metrics listener with a bare router (`crates/phase-server/src/main.rs:2295-2307`), not through `build_router`, and scraped in-cluster by Prometheus via PodMonitor — bypasses Traefik and the CompressionLayer entirely. Internal-only and modest; scope note, not a defect.

**No other internal comms to cover:** no pod↔pod traffic exists (server-core/seat-reducer/lobby-broker have no HTTP client; lobby-broker is in-process; scale-out ordinals are each Traefik→per-ordinal-Service under the same middleware chains), and the rendered workloads run a single container — no sidecars.

**Caveat:** the driven measurement used Traefik v3.5; the cluster's Traefik version is not pinned anywhere in this repo (the chart consumes an existing `traefik.io/v1alpha1` install). The forwarding behavior is standard Go-proxy semantics and unchanged in the v2/v3 compress sources read, but only v3.5 was exercised.

**Comment by lgray** [comment]
:robot: _AI text below_ :robot:

**Maintainer position (relayed on behalf of @lgray):** this PR is not sufficient on its own — the WebSocket traffic it leaves untouched is the bulk of real bandwidth, and completing that half is expected of this PR's author.

The measured picture (details in the two review comments above): the origin `CompressionLayer` + Traefik middleware work correctly, including on the cluster-internal hop — but they only ever see the small HTTP surface (`/health`, the backup API, `/admin`). `/ws` is the sole realtime channel, so effectively all game/lobby/draft bandwidth remains uncompressed on every hop, and no HTTP-layer middleware can ever reach it post-upgrade.

The deferral of lever 3 remains valid — re-checked today: tungstenite 0.30.0 (latest, July 2026) still ships no permessage-deflate, so the WS-stack route stays closed. That leaves **lever 4, the app-level gzip envelope tracked in #8274, as the only live route**, and the maintainer's expectation is that the author of this PR takes on #8274 as the follow-through: binary-frame handling on both ends, a capability flag in the `ClientHello`/`ServerHello` handshake so a compressed frame never reaches a peer that can't decode it, and per-client encoding at broadcast fan-out for mixed-version rollouts — as already scoped in that issue.

One sequencing note to save you a painful rebase: the same files (`server-core` protocol, `phase-server` WS dispatch) are being reworked on an in-flight branch that will land as a single large PR shortly. Starting #8274 after that merge avoids a guaranteed conflict.

Maintainer sign-off on the compression work as a whole is withheld until the WS half is addressed.

**Comment by lgray** [comment]
@matthewevans re-review considering my most recent commentary.

**Review by ntindle — DISMISSED** [review]
## hard_stop (pr-review-loop)

**Head:** `1c335312e5bee5c33fa00b83fdfe7040b3c46602`  
**Verdict:** REQUEST_CHANGES

This PR touches a hard-stop path from `.agents/pr-review-policy.toml`:

- `.github/workflows/helm-chart.yml` (matches `.github/workflows/**`)

Workflow / CI config changes are maintainer-owned and cannot ship through the normal contributor review loop. Please drop the workflow edit from this PR (split deploy/app compression work into a PR without `.github/workflows/**`, or get an explicit maintainer exception with the `accepted` path if this must stay coupled).

Prior CHANGES_REQUESTED on an older head does not cover this head — re-requested against `1c33531…`.

**Review by matthewevans — CHANGES_REQUESTED** [review]
**[MED] Explicit Traefik priorities make the general HTTP route globally losable.** Evidence: `deploy/helm/phase-server/templates/ingressroute.yaml:46-67` and `:108-120` set `/ws` to `100` and the overlapping `Host()` route to `1`; Traefik uses an explicit `priority` instead of its normal rule-length ordering. Why it matters: on a shared entrypoint, any overlapping router with priority above `1` can capture ordinary traffic for this host, whereas the previous `Host()` route retained its default priority. The chart test at `deploy/helm/phase-server/tests/assert-compression-boundary.sh:37-45` verifies middleware membership only, so it cannot catch that regression. Suggested fix: remove the paired explicit priorities and assert the rendered default-priority contract, or document and test ranks demonstrated safe against the complete operational router set.

**Comment by ITC-MSM** [comment]
@ntindle Addressed the hard-stop at head `1c335312e`.

`.github/workflows/helm-chart.yml` is reverted to exactly its `upstream/main` state at head `0665950` — `git diff upstream/main...HEAD -- .github/workflows/helm-chart.yml` is empty. The CI wiring for the compression-boundary assertion is dropped from this PR entirely; `deploy/helm/phase-server/tests/assert-compression-boundary.sh` stays as a standalone, contributor-owned test script that a maintainer can wire into the workflow themselves if/when they choose, per the policy's suggested path.

(Note for context: the workflow edit at `1c335312e` was itself a maintainer-authored commit pushed directly to this branch — see matthewevans' comment above — but per the policy this still routes through the hard-stop, so it's reverted regardless of who added it.)

**Comment by ITC-MSM** [comment]
@matthewevans Addressed the explicit-priority finding at head `06659509f`.

Verified against Traefik's routing docs (routers/rules-and-priority reference): when a router has no explicit `priority`, Traefik sorts routers **on the same entrypoint** by descending rule-string length, and `priority` values are scoped to the whole entrypoint — not to any particular pair of routers. That's exactly the fragility you flagged: pinning the general route to `priority: 1` made it globally losable to *any other router on the entrypoint* that keeps Traefik's (near-universal) default, which is a much larger blast radius than the two-route ordering problem the explicit priorities were trying to solve.

Went with your first suggestion — removed the paired explicit priorities entirely — because the default rule-length ordering is provably safe here, not just empirically: the `/ws` rule's match string (`Host(...) && PathPrefix(`/ws`)`) is always the general rule's match string (`Host(...)`) plus a fixed suffix, so it is unconditionally longer, for every possible host value. That structural property (not just today's rendered YAML) is what makes default ordering reliable long-term, so no future host/values change can invert it.

Changes in `deploy/helm/phase-server/templates/ingressroute.yaml`:
- Removed all four explicit `priority:` fields (entry route pair + per-ordinal route pair).
- Rewrote the explanatory comment to document why default rule-length ordering is sufficient and why explicit priorities here would be a worse trade (entrypoint-wide blast radius).

Changes in `deploy/helm/phase-server/tests/assert-compression-boundary.sh`:
- Added a regression assertion that fails if either rule in the pair pins an explicit `priority:` again.
- Added a structural check that the `/ws` rule's rendered `match:` string equals the general rule's `match:` string plus the exact ` && PathPrefix(`/ws`)` suffix, which is the property the whole safety argument rests on.

I hand-verified the new assertion actually catches a regression (rendered a scratch copy of the chart with `priority: 1` reintroduced on the per-ordinal general rule and confirmed `assert-compression-boundary.sh` fails on it) before landing this.

**Review by matthewevans — APPROVED** [review]
## Current-head approval

Reviewed at `6ecb858302068b8287f8dbec3fd6eb152de616b5`.

The current head restores the missing Helm CI invocation after the prerequisite default and canonical scale-out renders. The existing assertion exercises the compression boundary on both manifest shapes: `/ws` stays compression-free while the corresponding HTTP routes retain the middleware, including an ordinal route. The preceding priority fix remains intact; no explicit Traefik route priority is rendered.

No unresolved implementation finding remains. The Helm executable is unavailable in this worktree, so CI is the required fresh execution environment for the rendered-manifest assertion.
